sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/TableCatalog.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2;
+
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public interface TableCatalog extends CatalogProvider {
```

Comment:
The semantics here aren't clear at least to me. Typically a `<something>Provider` is a class that can instantiate a `something`. Here it appears to provide itself? The abstraction I would imagine would be to have either:

* `CatalogProvider` has a method called `get(options, sqlConf)` which returns a `TableCatalog` configured with the given options and `SQLConf`, or
* Remove `CatalogProvider` entirely and put `initialize` in this interface. Every `TableCatalog` instance must be initialized before calling any other methods like `loadTable`.

Reply:
The intent is to use some interface to load all catalogs, whether they implement `TableCatalog`, `FunctionCatalog`, or both (or other catalog API parts). So you load a catalog, then check whether it is a `TableCatalog` when you want to use it for tables.

Sounds like the name `CatalogProvider` is the confusing part. You're right that a provider usually implements a get method to provide something. I could change that to `CatalogImpl` or something. Would that work?

Reply:
Perhaps something as simple as `Catalog` as the top level then, which is sub-interfaced by `TableCatalog`, `FunctionCatalog`, and other "kinds" of catalogs. They can all share the same `initialize` method which is defined by `Catalog`. That sounds like the simplest idea, perhaps?

Reply:
What about `CatalogPlugin`? I'm hesitant to go with just `Catalog` because it isn't very specific. I think it might cause confusion because the interface has only the `initialize` method.

Reply:
I'm ok with that!